cmd/prometheus/main.go

```diff
@@ -282,104 +277,193 @@ func main() {
+	// early as possible, but ignore it until we are ready to handle reloading
+	// our config.
+	hupReady := make(chan bool)
 
-	go func() {
-		defer close(dbOpen)
-
-		level.Info(logger).Log("msg", "Starting TSDB")
-
-		db, err := tsdb.Open(
-			cfg.localStoragePath,
-			log.With(logger, "component", "tsdb"),
-			prometheus.DefaultRegisterer,
-			&cfg.tsdb,
+	var g group.Group
+	{
+		term := make(chan os.Signal)
+		signal.Notify(term, os.Interrupt, syscall.SIGTERM)
+		cancel := make(chan struct{})
+		g.Add(
+			func() error {
+				select {
+				case <-term:
+					level.Warn(logger).Log("msg", "Received SIGTERM, exiting gracefully...")
+				case <-webHandler.Quit():
+					level.Warn(logger).Log("msg", "Received termination request via web service, exiting gracefully...")
+				case <-cancel:
+					break
+				}
+				return nil
+			},
+			func(err error) {
+				close(cancel)
+			},
 		)
-		if err != nil {
-			level.Error(logger).Log("msg", "Opening storage failed", "err", err)
-			os.Exit(1)
-		}
-		level.Info(logger).Log("msg", "TSDB started")
-
-		localStorage.Set(db)
-	}()
-
-	prometheus.MustRegister(configSuccess)
-	prometheus.MustRegister(configSuccessTime)
-
-	// The notifier is a dependency of the rule manager. It has to be
-	// started before and torn down afterwards.
-	go notifier.Run()
-	defer notifier.Stop()
-
-	go ruleManager.Run()
-	defer ruleManager.Stop()
-
-	go targetManager.Run()
-	defer targetManager.Stop()
-
-	// Shutting down the query engine before the rule manager will cause pending queries
-	// to be canceled and ensures a quick shutdown of the rule manager.
-	defer cancelCtx()
-
-	errc := make(chan error)
-	go func() { errc <- webHandler.Run(ctx) }()
+	}
+	{
+		// Make sure that sighup handler is registered with a redirect to the channel before the potentially
+		// long and synchronous tsdb init.
+		hup := make(chan os.Signal)
+		signal.Notify(hup, syscall.SIGHUP)
+
+		cancel := make(chan struct{})
+		g.Add(
+			func() error {
+
+				select {
+				case <-hupReady:
+					break
+				// in case a shutdown is initiated before the hupReady is released
+				case <-cancel:
+					return nil
+				}
 
-	<-dbOpen
+				for {
+					select {
+					case <-hup:
+						if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
+							level.Error(logger).Log("msg", "Error reloading config", "err", err)
+						}
+					case rc := <-webHandler.Reload():
+						if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
+							level.Error(logger).Log("msg", "Error reloading config", "err", err)
+							rc <- err
+						} else {
+							rc <- nil
+						}
+					case <-cancel:
+						return nil
+					}
+				}
 
-	if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
-		level.Error(logger).Log("msg", "Error loading config", "err", err)
-		os.Exit(1)
+			},
+			func(err error) {
+				close(cancel)
+			},
+		)
 	}
+	{
 
```

Comment:
Nit: The usage of empty lines is inconsistent across this code. I'd vote to not add empty lines after `{` consistently.